### PR TITLE
[CDRIVER-2407] Android NDK compatibility

### PR DESCRIPTION
**Pull request by promethe42**
Fix/implementation for https://jira.mongodb.org/browse/CDRIVER-2407

### CODE REVIEW

**Comment by ajdavis** [comment]
Thanks for this, I'm sure it's very useful for Android and other constrained platforms.

However, this is a bigger change than I think is required. libbson only uses rand_r in one function, `_bson_context_init()`, which is used to generate random parts of ObjectIds. I propose a different approach with less potential for unintended consequences that will still support ObjectIds on Android.

Does Android have the thread-unsafe `rand()`? If so, then I suggest you do this:

- Keep the new checks in CMakeLists.txt and FindDependencies.m4, those are great
- Delete the new .h and .c files and `rand_r` implementation from musl
- Update `_bson_context_init()`: right now it checks `BSON_OS_WIN32` and tries to use `rand_r` otherwise. Instead it should check your new symbol `BSON_HAVE_RAND_R`.

That should suffice for Android if `rand()` is available there. The need for thread-safe `rand_r` is very rare, it's only if your own application code is, for some reason, calling [bson_context_new|http://mongoc.org/libbson/current/bson_context_new.html] concurrently in several threads. I don't know why an application would call `bson_context_new()` at all, to be honest: by default libbson initializes its BSON context the first time it creates an ObjectId with no special code in your application at all.

**Comment by promethe42** [comment]
@ajdavis like this ?

```c
#ifdef BSON_OS_WIN32
   /* ms's runtime is multithreaded by default, so no rand_r */
   srand (real_seed);
   context->seq32 = rand () & 0x007FFFF0;
#elif defined(__FreeBSD__) || defined(__NetBSD__) || defined(__DragonFly__) || \
   defined(__OpenBSD__)
   arc4random_buf (&context->seq32, sizeof (context->seq32));
   context->seq32 &= 0x007FFFF0;
#elif defined(BSON_HAS_RAND_R)
   context->seq32 = rand_r (&real_seed) & 0x007FFFF0;
#else
   context->seq32 = rand () & 0x007FFFF0;
#endif
```

**Comment by ajdavis** [comment]
Great, that looks right to me too.

File: CMakeLists.txt [hunk 1]
```diff
@@ -128,7 +135,7 @@ else ()
    endif ()
    set (BSON_EXTRA_ALIGN 1)
    CHECK_SYMBOL_EXISTS(SYS_gettid sys/syscall.h BSON_HAVE_SYSCALL_TID)
-   if (NOT BSON_HAVE_SYSCALL_TID OR APPLE)
+   if (NOT BSON_HAVE_SYSCALL_TID OR APPLE OR ANDROID)
```
Comment (ajdavis) [inline CMakeLists.txt]: Curious - was BSON_HAVE_SYSCALL_TID set to 1 on Android when it shouldn't be? You can debug with:
```
message(STATUS "BSON_HAVE_SYSCALL_TID = ${BSON_HAVE_SYSCALL_TID}")
```
I see that my colleague Hannes added the "OR APPLE" predicate with the commit comment, "build libbson on tvOS / watchOS". But it seems to me that CHECK_SYMBOL_EXISTS should do all the necessary checks itself without additional checks on this line?

Comment (promethe42) [reply CMakeLists.txt]: Fixed.
But I think there is a problem with Android NDK r16 and `BSON_HAVE_SYSCALL_TID`.

I'll assess this properly and do another PR for r16 if that's ok.

**Comment by ajdavis** [comment]
Thanks! After you investigate the `BSON_HAVE_SYSCALL_TID ` question, let me know if Android NDK compatibility is complete or if we need another patch.

**Comment by promethe42** [comment]
I still have build issues with `rand_r` on Android r16.
I'll keep you posted.

**Pull request merged**